### ISSUE TITLE
Distorted audio playback with sf::Sound and sf::Music

### ISSUE THREAD

**Issue by EichenseerMarco** [SFML/SFML#1813]
## Subject of the issue

Playing the attached sine sweep (logarithmic 10 to 22kHz in 20 seconds) with either sf::Sound or sf::Music produces very obvious distortions in the medium to high frequency range.
The distortions start around 3kHz and become worse the higher the frequency. 
These distortions cannot be observed with any other app I tried so I don't think it is a hardware issue on my end. 
Tested audio players: Windows 10 Groove Music, VLC, Chrome's internal music-player, Rhythmbox (Ubuntu VM)

The same issue can be observed with a static tone. 
The SFML playback of the attached 6k.ogg also sounds different compared to other applications. 

Both example files were created using audacity with a sample rate of 44100 and the track format set to 16-big PCM. 
I also tried different settings as well as sine sweep files I found online but the issue persisted. 

## Your environment

* Windows 10 Version 20H2
* SFML 2.5.1 (I also tried the latest development snapshot of the DLLs)
* MSVC 16.9.0  Compiler Version 19.28.29910

Also reproduced on

* Ubuntu 20.4 Virtual Box VM 
* SFML 2.5.1
* GCC 9.3.0

## Steps to reproduce

Input Files:
[ReproduceFiles.zip](https://github.com/SFML/SFML/files/6944584/ReproduceFiles.zip)

```cpp
#include <SFML/Audio.hpp>

int main()
{
	sf::SoundBuffer buffer;
	buffer.loadFromFile("sweep.ogg");
	sf::Sound sound(buffer);

	sound.play();
	while (sound.getStatus() == sf::Sound::Status::Playing)
	{
		sf::sleep(sf::milliseconds(100));
	}

	return 0;
}
```

## Expected behavior

The sweep is played without distortions like in other applications.

## Actual behavior

Starting at around 3kHz very obvious sound distortions start to occur.

**Comment by eXpl0it3r** [comment]
For the sweep I can hear some "overtone" sweeping (faster sweeping within the sweep) in higher frequencies, which I wouldn't really call "distortions" though. Did you hear the same, or does it generate some other kind of distortion?

The 6k example is more interesting and trying to record this (with compression), it generates an interesting interleaved square signal, which might hint at some odd limiting:

![image](https://user-images.githubusercontent.com/920861/128494929-4ec806b6-57e0-45d6-a0c2-98c4df0f57ac.png)

Another thing I tried was using *.wav files instead of *.ogg and the sweep seems to be clean or cleaner, while I still hear some distortion for the 6k.

This is quite tricky to debug, since it could be in various places:
- SFML audio file reader
- Vorbis library not being used correctly
- Limit of how SFML stores and handles audio samples
- SFML not properly calling OpenAL
- OpenAL having some limitations

Maybe @LaurentGomila has some insights where to check first or known-limitions.

**Comment by EichenseerMarco** [comment]
> For the sweep I can hear some "overtone" sweeping (faster sweeping within the sweep) in higher frequencies, which I wouldn't really call "distortions" though. Did you hear the same, or does it generate some other kind of distortion?

I'm not really familiar with the audio terminology so I just called it "distortion". Faster sweeping within the sweep is exactly what I hear as well.

> Another thing I tried was using *.wav files instead of *.ogg and the sweep seems to be clean or cleaner, while I still hear some distortion for the 6k.

Using the same audacity settings but exporting to wav produces the same results as the ogg files.
I attached the both wav files just in case its something with my settings: [ReproduceFilesWav.zip](https://github.com/SFML/SFML/files/6944932/ReproduceFilesWav.zip)

**Comment by Hapaxia** [comment]
It's worth noting that opening the .ogg file in Audacity actually shows some visible inaccuracies that could be causing some of these artifacts. This may be why the .wav file can give cleaner results. It's also possible that some players are simply filtering certain frequencies for human comfort.

I've attached a screenshot of Audacity with the .ogg file you provided of the sweep - zoomed in - and you can see that it's not just a simple, single-frequency sound wave.
![sweep](https://i.imgur.com/rcKxdPj.png)

Even the 6k regular tone looks 'unpure' when opened directly in Audacity:
![6k](https://i.imgur.com/6aKhkEl.png)

Compare to a 'pure' 6k tone generated in Audacity and you can see that it's much more regular:
![6k pure](https://i.imgur.com/VcQZHYQ.png)

**Comment by EichenseerMarco** [comment]
I'm not sure what you mean with my sweep not being a "simple, single-frequency sound wave". A sweep can't be a single frequency. 
I checked your file and it also looks like that. How would a clean sweep look like?

![image](https://user-images.githubusercontent.com/20540864/142743674-4786bedd-c5f1-4a0c-a72f-757f769738cd.png)

&nbsp;

The impure looking 6k seems to be due to the different sampling rate. I used 44100, your files are 48000.
With 48000Hz a 6k tone look more regular.
I guess this is because 48000 is a multiple of 6000 so the sampling points always hit the same point in the sine wave making it look more regular.

&nbsp;

>  the unwanted sounds don't seem to be very audible when playing back the sweep even in Audacity

That is interestng, I can't hear anything strange when listining to my files in Audacity (or any other player). 
Did you have to do anything special to hear it?

&nbsp;

I tried your sweep in SFML and the echo or faster sweeping within the sweep does indeed not occur! 
This lead me to an interestng discovery.
The faster sweeping seems to occur only with a 44100kHz file. Your sweep is 48kHz, mine is again 44.1kHz.
I created another 20-20kHz sweep with Audacity but setting the project to 48kHz and the faster sweeping again does not occur. 

I initially also tested this with sweeps I found online to rule out bad Audacity settings but the files I used all were 44.1kHz as well. 
I found another sweep online that is 48kHz and even converting the ones I already have to 48kHz makes the faster sweeping disappear. 
(Its the same with the strange sounding 6k tone, using a 48kHz file sounds normal)

So this seems to indicate a bug related to 44.1kHz sampling rate playback.

**Comment by eXpl0it3r** [comment]
Can anyone test this with SFML 3 again?

Not sure if we want to invest more time into SFML 2 fixes.

**Comment by EichenseerMarco** [comment]
Hi,

I'm still able to reproduce the issue with SFML 3.0.1.

**Comment by EichenseerMarco** [comment]
I'm also able to reproduce the issue with the [high level API example](https://github.com/mackron/miniaudio?tab=readme-ov-file#examples) from the miniaudio repository. 

It's interesting that the same problem occurs with a different audio backend, but it doesn't appear to be an SFML bug after all.

**Comment by eXpl0it3r** [comment]
Could it be the same issue as with #3526 ?

**Comment by EichenseerMarco** [comment]
Yes, I think so. The 44100-48000.mp4 example sounds exactly like the problem I'm describing here. 

I wasn't sure what oomek meant with "audio interface is set to 48000Hz" but that lead me to the discovery that this windows setting has an effect. 
The problem disappears if this matches the sampling rate of the played file. 

![Image](https://github.com/user-attachments/assets/fdfd6a44-67b7-4ef5-ad1f-babb60a2c3cb)
